### ISSUE TITLE
RFC / RFE: LOH analysis in tumor-normal samples

### ISSUE THREAD

**Issue by lbeltrame** [bcbio/bcbio-nextgen#1876]
GIven the interest in studies that involve tumor heterogeneity / subclonality, currently bcbio offers "out of the box" support for both somatic variants and CNVs. A useful metric that can be combined (and already used by some tools, like CNVkit's plotting) is LOH, which (to my knowledge) is not yet handled.

I admit I'm not sure if there is support already for this in bcbio. I know that back in the days I baked VarScan support to actually *remove* LOH calls from the VCF as they weren't truly somatic calls. 

The biggest problem here is how to actually and reliably extract these information. MuTect[2] might have these in the REJECTed calls (but how to distinguish them?), VarScan 2 calls them (might just be needed to move them away and elsewhere) and I'm not sure how FreeBayes and VarDict handle them. 

Or are there any other tools more suited for this purpose? 

I'm willing to put the money where my mouth is in this case as we're starting to explore this in my institution and having bcbio do that would greatly streamline things.

**Comment by lbeltrame** [comment]
Some follow ups:

- VarDict adds `LikelyLOH` and `StrongLOH` to the `STATUS` line in the `INFO` field (although these are called `REJECT`);
- MuTect2 has no information, one might have to look at genotypes;

I don't have any FreeBayes data here, so I can't check.

**Comment by chapmanb** [comment]
Luca;
Thanks for starting this discussion. We've been continuing to work on plans for improving heterogeneity support in bcbio, which would include LOH detection. Here are my notes:

https://github.com/bcbio/bcbio.github.io/blob/master/_posts/2017-03-25-heterogenity.md

We currently have BubbleTree integrated which supports LOH calling, although it requires manual extraction from their generated plots. The rough work plan now is to include TitanCNA which should provide this explicitly, using PureCN as a pre-Titan step for tumor-only inputs.

Integrated copy number and small variant approaches should provide an improvement over variant only outputs and also feed into downstream heterogeneity estimation algorithms like PhyloWGS. Practially, VarScan and MuTect2 are also not licensed cleanly for non-academic use and given the complexity here it would be helpful to have a single well-supported approach. TitanCNA is now GPL licensed so could be the center of that approach as long as the integration validates well. 

I'd love to hear thoughts and suggestions and happy to share more as we work on implementing these.

**Comment by lbeltrame** [comment]
Thanks Brad, I'll make sure to go through the notes. Are there any bite-sized jobs that can be done in this area? I'd be happy to help.

**Comment by lbeltrame** [comment]
FTR: a major issue I've found with some of the tools that I've been researching (but not those you mentioned, I need to dig deeper) is that they are suited for WGS workloads. I'm mostly interested in something that can work at a WES level, because of cost and because these can also be done efficiently on smaller instruments like the NextSeq 500.

**Comment by schelhorn** [comment]
+1 for supporting WES and tumor-only as well - very relevant for us.

**Comment by lbeltrame** [comment]
Some further rumination: for use in downstream tools like PhyloWGS we need also a VCF file. Should we support for that only one or two callers, like VarDict and FreeBayes (also for cleaner licensing)?

I've briefly looked at TitanCNA, I might try to run it standalone to see how it fares. It's mainly meant to be used from R or their ruffus-based pipeline. To use it we may want to use either a R script baked for the purpose, or `rpy2` to use it from Python (caveat: I use that daily and works OK, with the disadvantage that's not always trivial how to map things from the R to Python space and vice versa)..

**Comment by lbeltrame** [comment]
Some more notes:

TitanCNA requires depth from REF and ALT in heterozygous loci identified in the normal sample, taken from the tumor sample (so get the AD in the tumor from the coordinates of a 0/1 locus in the normal). Other software packages want the same, as far as I can tell.

This means however that a complete "germline" run needs to be done for the pair, and not only for the normal. Any suggestions for an alternative approach?

**Comment by chapmanb** [comment]
Luca and Sven-Eric;
Thanks for all the thoughts and suggestions. I put together a short presentation that walks through the current status and proposed implementation that might be helpful to orient the rough notes I sent earlier:

https://github.com/chapmanb/bcbb/blob/master/talks/heterogeneity2017/heterogeneity2017_bcbio.pdf

This should give a path to WES and also tumor-only as long as you have process matched normals.

Luca, I don't have any practical TitanCNA experience yet to answer your questions in depth but in general for a tumor/normal pair we'd call both together giving us somatic and filtered germline calls, and then also call normal germline only. Then we can use the counts from the somatic/normal calls based on hets in the germline. Tumor-only is trickier, and we'd need to use PureCN to help classify as tumor/germline for a preliminary step. 

I'd call TitanCNA standalone via R and think we can probably parse the inputs we need from current bcbio generated files, rather than using their prep scripts. Once we get into some of this process implementation I'd probably have smaller size chunks of work to suggestion but only have this broad plan now. Helpful things to do immediately would be organizing validation datasets so we can ensure we have correct and useful outputs once we put these in place.

Thanks again for all the discussion.

**Comment by lbeltrame** [comment]
Thanks Brad. I went through the notes and indeed there's a long journey ahead of us, but let's go one step at a time and we'll get there. ;) 

Currently I'm experimenting with TitanCNA to get a feel on how it works, and what output produces. My next step is to jot down notes on how the whole process works (and not forgetting to read a few papers on the topic) and come up with at least a PoC (likely standalone). Once this is in place, we can look at validation data sets (the Illumina link you gave is helpful).

**Comment by lbeltrame** [comment]
More notes. 

- TitanCNA generates read count bins based on 1000bp non-overlapping windows, by default. Do we have something that does the same already? (Would save installing their programs like `readCounter`);
- `gcCounter` (which is used with the FASTA reference) calculates GC content over 1000bp non-overlapping windows. Precomputed values would likely be possible for the genomes shipped with bcbio
- `mapCounter` does the same on mappability regions from UCSC (Again, this can be a one-time affair)
- Read binning needs to be done twice, once per normal, once per tumor, The HMMcopy docs tell that these files should have an identical number of bins
- All pre-analysis files are wiggle files, with the exception of the calls, that are a table with: chromosome, position of 0/1 call, reference base, reference depth in tumor, alt base, alt base depth
- All these utilties are part of `HMMcopy` rather than TitanCNA (also GPLv3).

**Comment by chapmanb** [comment]
Luca -- thanks for the notes. My idea is to use CNVkit for the read counts, segmentation and GC correction inputs. @etal has been awesome at integrating CNVkit with other downstream tools so hopefully this is a do-able target. I'd love to hear his thoughts in general on this discussion.

**Comment by etal** [comment]
I hadn't noticed that TitanCNA is now GPL3, that's great. From Luca's notes it sounds like TitanCNA has a modular design like CNVkit, so surely some files can be reused.

For WGS, goleft could also provide binned read counts and GC. Some light scripting might be needed to convert those to TitanCNA's input formats. I'm not sure how the TitanCNA/HMMcopy pipeline works for exomes or target panels; maybe CNVkit's binning and GC would be useful there.

For unpaired tumor samples, CNVkit can/does in some cases simulate normal-sample read depths from each bin's weight (e.g. `export theta`). This is generally better than nothing, and the code is modular, so maybe this and other CNVkit functionality will be useful for getting around edge cases for TitanCNA and the other heterogenity/LOH tools.

I've started work on a 'clonality' command (etal/cnvkit#200) meant to streamline interop with these other tools.

If tumor purity is known, CNVkit will currently output allele-specific integer copy number for each segment, based on total copy number and b-allele frequency in a given SNV VCF. That's not the same as identifying per-SNV LOH, which I thought this thread started with, or runs of homozogosity, which I suppose are more of a concern for germline/population analysis. But I'm happy to help support any of this if there's community interest.

**Comment by lbeltrame** [comment]
Thanks for all the suggestions. Where I can download a single chromosome BAM from the TCGA-DREAM challenge? I'd use that to make a wiggle file with the tools from TitanCNA, so that @etal and possibly anyone other interested can look through the output.

I'll also post the genome GC wiggle files and the mappability scores so that those can be looked at as well. In the mean time, I have done a single tumor/normal run via the GATK and I'll see if I can write a PoC to convert things (o/1 calls) in the format TitanCNA wants.

What VCF parsing library do we use in bcbio currently? PyVCF, pysam, or cyvcf2?

**Comment by lbeltrame** [comment]
I did some PoC code: on a final VCF made by bcbio, this produces the output needed for TitanCNA (it's ugly, because I knew in advance what was the position of the normal sample in the two).

I do a tuple because I'll feed it in a pandas DataFrame for my experiments, but I can easily just write it directly to file.

```python
import cyvcf2
positions = list()
reader = cyvcf2.Reader("TitanCNA-gatk-haplotype-annotated.vcf.gz")
for record in reader:
    
    if record.FILTER or not record.is_snp:
        continue

    # Normal sample genotype
    genotype_type = record.gt_types[1]
    
    # Not HET, skip
    if genotype_type != 1:
        continue
    
    # From tumor sample
    alt_depth = record.gt_alt_depths[0]
    ref_depth = record.gt_ref_depths[0]
    
    positions.append((record.CHROM, record.POS, record.REF, ref_depth, 
                      record.ALT[0], alt_depth))
```

**Comment by lbeltrame** [comment]
More notes on my tests: the computation would need to be run from 1 to 5 times, because getting the right number of clones requires running it with different subclonal population numbers (from 1 to 5) and then selecting the one with the lowest statistic.

My idea would be to run it from 1 to 5 and let the user decide in the end (but that would work when feeding things to PhyloWGS?). @chapmanb @etal Opinions?

**Comment by chapmanb** [comment]
Luca;
Thanks for digging into this and the initial scripts for converting the VCF outputs. For TitanCNA, the best approach is to run a single time and then let downstream tools like PhyloWGS handle the subclonal population analysis. It would be useful to understand how stable the TitanCNA outputs are between different subclonal numbers: do CNV allele placements and LOH stay relatively stable on different outputs? We'd like to avoid running multiple times and making it that complex unless it provides much more reliable outputs. Thanks again.

**Comment by lbeltrame** [comment]
That's my next step. I'll check the comparisons wwith 1, 2, 3, 4, and 5 clones on my initial two sample test and see how different they are. In the mean time I'm getting some germline VCFs done to see how the whole process can be automated.

**Comment by lbeltrame** [comment]
In this limited example with 2 samples only, the *call* (so LOH, or loss, or generic CNA) made by TITAN is constant across different clone numbers from 1 to 5:

```python
In [50]: clone_state.head()
Out[50]: 
             titancna_cloneno.1 titancna_cloneno.2 titancna_cloneno.3  \
Chr Position                                                            
1   909238                 NLOH               NLOH               NLOH   
    986787                 NLOH               NLOH               NLOH   
    1026310                NLOH               NLOH               NLOH   
    1156131                NLOH               NLOH               NLOH   
    1387764                NLOH               NLOH               NLOH   

             titancna_cloneno.4 titancna_cloneno.5  
Chr Position                                        
1   909238                 NLOH               NLOH  
    986787                 NLOH               NLOH  
    1026310                NLOH               NLOH  
    1156131                NLOH               NLOH  
    1387764                NLOH               NLOH 

In [51]: clone_state.eq(clone_state.iloc[:, 0], axis=0).all(1).value_counts()
Out[51]: 
True     16684
False     2591
dtype: int64
```

Which means that the 86% of the positions called have the same call independently from the number of clones selected in the analysis (sample size of 2, first test, etc. disclaimers apply). The identified positions are always the same:

```python
In [54]: clone_state.isnull().any()
Out[54]: 
titancna_cloneno.1    False
titancna_cloneno.2    False
titancna_cloneno.3    False
titancna_cloneno.4    False
titancna_cloneno.5    False
dtype: bool
```

**Comment by lbeltrame** [comment]
And of course, how I did it:

```r

require(TitanCNA)

# Tumor depth, binned
tumor_wig = tumor.wig"
# Normal depth, binned
normal_wig = "normal.wig"
# GC content, binned
gc_wig = "gc.wig"
# Mappability. binned
map_wig = "map.wig"

cn_data = correctReadDepth(tumor_wig, normal_wig, gc_wig, map_wig)

compute_titancna = function(filename, cn_data, clusters=2, cores=20) {
    require(doMC)
    data = loadAlleleCounts(filename, genomeStyle="NCBI")
    params = loadDefaultParameters(numberClonalClusters=clusters, data=data)
    logR <- getPositionOverlap(data$chr, data$posn, cn_data)
    data$logR <- log(2^logR)
    data <- filterData(data, c(1:22, "X", "Y"), minDepth = 10, maxDepth = 200, positionList = NULL, 
        centromere = NULL, centromere.flankLength = 10000)
    registerDoMC(cores = cores)
    convergeParams <- runEMclonalCN(data, gParams = params$genotypeParams, nParams = params$normalParams, 
        pParams = params$ploidyParams, sParams = params$cellPrevParams, 
        maxiter=20, maxiterUpdate=1500, useOutlierState = FALSE, txnExpLen = 1e15,  
        txnZstrength = 5e5, normalEstimateMethod = "map",  estimateS = TRUE, 
        estimatePloidy = TRUE)
    optimalPath <- viterbiClonalCN(data, convergeParams)
    filename = paste("~/titancna_cluster.", item, ".txt", sep="")
    paramfile = paste("~/titancna_parameters_cluster.", item, ".txt", sep="")
    results <- outputTitanResults(data, convergeParams, optimalPath,  
        posteriorProbs = FALSE,  subcloneProfiles = TRUE, filename=filename)
    outputModelParameters(convergeParams, results, paramfile,  S_Dbw.scale = 1)
  }

for (item in seq(1,5)) {
    cat(paste("Number of clones", item, "\n"))
    compute_titancna("~/TitanCNA_data.txt", cn_data, clusters=item)
 }
```

**Comment by chapmanb** [comment]
Markus -- thanks so much for the details on the additional PureCN work you're doing; CNVkit support will be incredibly helpful for us. We really appreciate you making PureCN available as we end up with a lot of similar samples -- poor quality tumor-only panels -- that we'd like to be able to diagnose as much as possible. From the bcbio side, I'll work to get PureCN included along with TitanCNA and we start to do some practical side-by-side comparisons.

One of the things I'm not yet sure how to get right is selection of a process matched normal. What do you think about the approaches in the recent SynthEx paper:

https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1193-3

Ideally we could provide an automated method to select the best normal from a set of choices, and have some estimates about how reliable the CNV/purity/clonality estimates relative to that choice.

Thanks again for PureCN and the helpful discussion.

**Comment by lima1** [comment]
Good question. I saw that paper and will benchmark SynthEx at one of the next weekends.  At first glance, it looked similar to what people have routinely done in the past - not being dismissive of their work, nobody so far succeeded in putting all the many good ideas of all major tools perfectly together.  

I like some of the details and it might be better than our default (hierarchical clustering vs PCA; minimizing MSSD vs Euclidean distance, weighting clusters instead of using only the k best samples).  In my experience, even something very simple like picking the best normal produces good results for most samples with little impact on purity/ploidy. Incorporating the variance information in the pool seems to remove some additional noise due to low quality regions and germline CNAs (as it of course should). GATK4's kind of related approach also makes a lot of sense to me (https://github.com/broadinstitute/gatk-protected/blob/master/docs/CNVs/CNV-methods.pdf)

I was surprised to see how well CNVkit works without pre-selecting normals at all. I sometimes get slightly cleaner log ratios with out default in exomes (not in small panels without tiling probes where off-target reads help a lot), but haven't looked into those samples thoroughly yet. We are for example more aggressive in the segmentation, which could explain some of the differences, but which probably also means we are in general not as sensitive as CNVkit.

Regarding CNVkit/PureCN integeration: upcoming v1.6 takes the DNAcopy style segmentation as input. I will experiment with incorporating CNVkit's coverage/log-ratio files as well (so that PureCN has access to the target level data, not only segment log-ratios). If this improves calling, it should make it in the developer version soon.

**Comment by etal** [comment]
For integration with PureCN and maybe others, CNVkit can export .cns files to the DNAcopy format with `cnvkit.py export seg`, and import the same with `cnvkit.py import-seg` -- hopefully saving some scripting. The `call` command also has more generic functionality such as calculating average SNP b-allele frequencies for each bin in a .cnr file, given a VCF. If you see any other missing functionality that would help integrate CNVkit with other programs like PureCN, feel free to suggest here: etal/cnvkit#200

There is some earlier literature I can point to for selecting a subset of normals from a pool:

- [CANOES](https://doi.org/10.1093/nar/gku345)
- [ExomeDepth](https://doi.org/10.1093/bioinformatics/bts526)

For clinical testing, it complicates traceability to allow a different set of normals to be selected for comparing to each tumor sample that comes through the pipeline, so we instead just build a single pooled reference from about 30 normals prepared and sequenced with the same protocol. Plus, I just haven't implemented that feature yet. But there is an open ticket for it: etal/cnvkit#126

In CNVkit's .cnr files I've tried to retain all the information that would be useful from a paired normal or pooled reference (absolute read depth, normalized log2 ratios, weight proportional to estimated spread, optional BAF), so e.g. `export theta` can generate read counts for a hypothetical paired normal from just the tumor .cnr. So that may be some help elsewhere, too.

**Comment by lbeltrame** [comment]
I almost finished preparing an internal dataset for TitanCNA analysis. I'll then batch the identified alterations by clone number and see how much variation occurs when considering differing clonal populations.

**Comment by lbeltrame** [comment]
BTW, semi-OT question: will any of you be at ECCB/ISMB 2017 in Prague? It would be nice to have a discussion on this in person.

**Comment by lbeltrame** [comment]
More notes: it would be good to do the binning ourselves (or with cnvikit) because I've found at least one BAM file that makes HMMcopy's readCounter crash. I haven't been able to pinpoint the issue.

**Comment by lbeltrame** [comment]
Additional input: TitanCNA 1.10 or newer is required, or it won't be able to export data in a format PhyloWGS understands.

**Comment by lbeltrame** [comment]
This paper (http://www.nature.com/ng/journal/v49/n5/full/ng.3838.html#methods) describes (very succintly) a possible strategy to find the optimal parameter combination in TitanCNA, but I worry that similar approaches might be very slow.

**Comment by lbeltrame** [comment]
After almost an hour spent on PhyloWGS parser, I can tell that TitanCNA output and our variant caller output needs *extensive* adjustments in order to satisfy its requirements (and not meeting them ends up in errors everywhere).

**Comment by chapmanb** [comment]
Luca -- I just wanted to say thank you for all this work. I apologize for not having yet had time to work on this myself. We're re-doing a lot of bcbio internals with CWL and I've been focusing on that. I'll hope to have more time to work on this once we clear that all out. Thanks again.

**Comment by lbeltrame** [comment]
No problem, I'm still very far from getting usable results.

One thing that I've noticed is that PhyloWGS Is **slow**. Granted, I have a large number of loci I used, but currently we're close to 24h run (for 2 samples together) and if I understood its output correctly, we're at 50% of the total analysis.

The real problem is that PhyloWGS (like EXPANDS, which I'm also testing) runs single core.

**Comment by lbeltrame** [comment]
"Slow" is an understatement. The most recent version of PhyloWGS has been running (with 2 samples) for *three days* straight, and it's not even remotely close to finishing (I would say it's obvious, since it runs single core). I'm not sure how this can be sped up.

**Comment by lbeltrame** [comment]
I understand I may have done things horribly wrong... but my analysis with PhyloWGS has taken almost a week running and it's still not done. 
Did anyone else try PhyloWGS out?

**Comment by lbeltrame** [comment]
Once the run finishes, I'll file an issue on their GH tracker to see what could be done (I'm sure I did something wrong).

**Comment by gavinha** [comment]
Hi, 

I was referred to this issue by @chapmanb and this is a great discussion.  I have some comments which I hope will be helpful.

 **Input data**
TitanCNA can accept allelic read counts and corrected log ratio coverage data from any source.  Recently, there have been some useful tools that can generate these input files. 
1. The Shah Lab at BC Cancer Agency has written a TITAN-specific workflow [KRONOS](https://github.com/MO-BCCRC/titan_workflow) that helps generate the data.  In brief, it uses HMMcopy to generate the log ratios and mutationSeq to identify germline HETs.   
2. The GATK team is also working on a tool that will generate these input files.  GATK-ACNV tool generates the TitanCNA input files and it is in the latest release.
https://github.com/broadinstitute/gatk-protected/releases/tag/1.0.0.0-alpha1.2.6 
```java -jar gatk-protected.jar ConvertACNVResults```

**Running TitanCNA**
I have provided scripts that I use myself, which includes updated default parameter settings, running the EM-training, post-processing to remove empty clonal clusters, adjusting final parameters, and plotting.  This is more for convenience but may be helpful for including plotting and post-processing steps.  
This script requires the input files as mentioned above.
https://github.com/gavinha/TitanCNA/tree/10X_Genomics/scripts
It is currently in a branch as part of some other extensions I'm working on.  I plan to merge this to the master and release as v1.15.0 soon.  As you can tell, I'm making extensions to support haplotype phasing and molecule-based coverage from 10X Genomics WGS.

**Selecting optimal solutions**
I know many users have had problems selecting the optimal solution and often, for me, it comes down to manual inspection.  The hardest step is selecting the correct ploidy solutions in an automated way.  I have a quick script (unfortunately for you, in R) that tries to automate a few of my manual inspection steps.
 https://github.com/gavinha/TitanCNA/tree/10X_Genomics/scripts/selectSolution.R

This script will accept the directories storing TitanCNA solutions for specific ploidy initializations and attempt to find the optimal ploidy based the consensus of the larger log-likelihood (which is already computed and in the parameter files) between ploidy solutions across all clusters. Then, it will select the optimal number of clusters, using the minimum S_Dbw validity index, only from the set of runs for the chosen ploidy initialization.

**Comment by lbeltrame** [comment]
Hello Gavin,

first of all, thanks for jumping in the discussion! 

> TitanCNA can accept allelic read counts and corrected log ratio coverage data from any source. Recently, there have been some useful tools that can generate these input files.

The requirements in the documentation are laid out well, so I anticipate this won't be an issue in bcbio (I was able to get this out of MuTect 2 fairly easily).

> I have provided scripts that I use myself, which includes updated default parameter settings, running the EM-training, post-processing to remove empty clonal clusters, adjusting final parameters, and plotting. This is more for convenience but may be helpful for including plotting and post-processing steps.

Are these optimized for WGS, or should be applicable to WES, too? I'm focusing on the latter since with the availability of instruments like the NextSeq 500, WES runs are feasible also for smaller institutions (like mine). I also ask because WES data is what I could test these approaches with.

>  The hardest step is selecting the correct ploidy solutions in an automated way. I have a quick script (unfortunately for you, in R) that tries to automate a few of my manual inspection steps.

That would be a good start. As you may have noticed from the paper I pointed earlier in the discussion, people have tried many different ways to handle the ploidy and clonality.

**Comment by lbeltrame** [comment]
It looks like the PhyloWGS slowness is due to processing multiple samples together. If using a single sample, each iteration takes about 10 seconds (as opposed to 3 minutes earlier).

**Comment by lbeltrame** [comment]
Leaving heterogeneity aside for a moment, I think it would be worthwhile to have a way to extract LOH from samples. [This paper](https://www.nature.com/articles/s41467-017-00388-9) is a good indication on why it would be worthwhile to pursue. 

I'm aware of VarScan 2 being used for this, and Sequenza (which I'm both testing at the moment).

**Comment by chapmanb** [comment]
Luca -- I agree about the importance of LOH and was hoping to get this as part of a more integrate package like TitanCNA and PureCN, rather than exploring additional tools. Part of this is just to reduce the complexity of multiple tools, and part is to be able to have an internally consistent set of calls for all these metrics which can help with validation since we lack reference standards. Sorry again I've been slow at allocating time to integrate and do more work on these and am still planning to do so.

**Comment by lima1** [comment]
I can recommend Sequenza at least for benchmarking (not citing it was a major, accidental omission in our PureCN manuscript). It's very easy to run and gives good results even with low purities and across all ploidy ranges.  But we weren't very lucky in getting it to work robustly without matched normals and on cell lines.

**Comment by lbeltrame** [comment]
TItanCNA has finally its 10X genomics branch merged, which means that the script mentioned here earlier is now usable. The update to R 3.3.2 also is in, so this is about the ideal time.  I would, as mentioned earlier, rely however on the tools that we have already to do the binning and the creation of the wiggle files that are needed.

**Comment by lbeltrame** [comment]
Paging again @etal  referring to my comment here (https://github.com/chapmanb/bcbio-nextgen/issues/1876#issuecomment-290037259): can CNVkit at this stage be used for the binning as explained in that comment?

**Comment by etal** [comment]
@lbeltrame Yes, CNVkit's targetcoverage.cnn files contain a "depth" column in absolute scale that should be easy to feed to other tools like TitanCNA. It uses htslib's bedcov for the calculation, so it should be as fast as anything other than `mosdepth`. But the bin sizes are not all the same, is that a problem?

**Comment by lbeltrame** [comment]
Yes, that is the problem, as far as I can see the bins used for the TitanCNA wigs are fixed (and need to be the exact same number for all inputs).

**Comment by lbeltrame** [comment]
Also, a possible adjustment towards selecting good genotypes (at least in some samples which may have artifacts):

```python
import cyvcf2
positions = list()
reader = cyvcf2.Reader("TitanCNA-gatk-haplotype-annotated.vcf.gz")
min_alt_depth = 8
for record in reader:
    
    if record.FILTER or not record.is_snp:
        continue

    # Normal sample genotype
    genotype_type = record.gt_types[1]
    
    # Not HET, skip
    if genotype_type != 1:
        continue
    
    # From tumor sample
    alt_depth = record.gt_alt_depths[0]
    ref_depth = record.gt_ref_depths[0]

   # Skip if too low, might get rid of artifacts
    if alt_depth < min_alt_depth:
        continue
    
    positions.append((record.CHROM.replace("chr", ""), record.POS, record.REF, ref_depth, 
                      record.ALT[0], alt_depth))

**Comment by gavinha** [comment]
Hi all,

Just want to quickly comment on input files to TitanCNA.  

The TitanCNA (module) can accept any read depth (log2) input, preferably GC-content bias corrected.  I typically use the HMMcopy approach to correcting read depth, which pre-dates TitanCNA.  I also continue to use it in other packages as well (e.g. [ichorCNA](https://github.com/broadinstitute/ichorCNA)).  I provide this as part of the TitanCNA workflow, but it can easily use alternatives.  

For example, in the main R script in the TitanCNA repo, [https://github.com/gavinha/TitanCNA/blob/master/scripts/R_scripts/titanCNA.R](https://github.com/gavinha/TitanCNA/blob/master/scripts/R_scripts/titanCNA.R), uses can pass in any read depth corrected data as long as it is in the tab-delimited format (`chr`,`start`,`end`,`logratio`).

I hope this clears things up and will help with re-using modules in your workflows.

Best,
Gavin

**Comment by lbeltrame** [comment]
Thanks, Gavin. @etal Given the above, would the data produced during the CNVkit processing be useful? It would save extra processing if so.

**Comment by lbeltrame** [comment]
Reading through the description in the docs, it looks like CNVkit's `cnr` files, which are already normalized and corrected wrt a reference, would do. I'll do some extra tests and will compare to the Titan-made ones.

**Comment by lbeltrame** [comment]
Dropping here (so I don't forget) a PoC code to convert a `cnr` file to the format Titan wants:

```python
import pandas as pd

# Easier on large files like cnr
iterator = pd.read_table(sourcefile, iterator=True)

with open(destination_file, "w") as handle:
    for chunk in iterator:
    chunk = chunk[["chromosome", "start", "end", "log2"]]
    chunk.columns = ["chrom", "start", "end", "logR"]
    chunk.to_csv(handle, mode="a", sep="\t", index=False)
```

**Comment by etal** [comment]
That code looks right to me. Note that the .cnr format uses 0-based indexing like UCSC BED, so if Titan expects 1-based indexing like R, then you need an extra line:

        chunk['start'] += 1

CNVkit will emit the same number of bins and bin coordinates in each .cnr file that was processed together, or with the same fixed reference.cnn. But the bin sizes vary. Looking at Gavin's script, I think this should be OK -- first the SNP allele frequencies are loaded into a dataframe, then the logR values are aligned to the SNP sites with a function that does not appear to require a fixed window size. The per-SNP logR values are then added as another column in the first dataframe, and then TitanCNA is done with the logR file.

**Comment by chapmanb** [comment]
Luca, Markus, Gavin and Eric;
Thanks for this incredibly helpful discussion. I've been re-working parts of coverage estimation in bcbio and am planning to centralize all the coverage based runs in a single location in the code to make this easier. Right now they happen in a variety of places: coverage calculation for parallelization, coverage for QC, and within individual SV callers to generate log2/cnr style files. This makes it hard to see overlaps and re-use between modules.

I'm starting to work on this as part of the 1.0.6 development cycle and then will probably have more specific questions. I'd like to use mosdepth for the actual depth calculations; we swapped it out for sambamba depth calculations and saw a ~3x speedup. I'll need to explore CNVkit, TitanCNA and PureCNs binning and GC approaches to be sure they're compatible but it sounds like the key is to have consistent bins across samples. We also have seq2c, which uses exon based binning, but I'd like to have as few strategies as possible or at least understand the tradeoffs of different bin approaches.

Thanks again for all this helpful work and discussion.

**Comment by lbeltrame** [comment]
I would say this is excellent timing. I was about to write that IMO TitanCNA is good to be used in the pipeline (modulo some small changes I'm going to submit PRs for). I would focus at first *only* on LOH, the rest can come later. THe only problem is that I didn't have time to find any good truth data set to run this on (I mainly focused on getting it to work properly).

**Comment by lima1** [comment]
PureCN uses the "readCoverageFile" function to parse all coverage data and this function currently supports GATK3 and CNVkit (*.cnr). If GC-normalization is performed by PureCN, it will write a new file that adds the necessary information whether a target is on- vs. off-target (although I could also add a heuristic that determines this based on interval size if this is a deal breaker). Happy to add any additional format you may need. Ideally the coverage output should contain both number of fragments and total coverage. 

I'm currently experimenting with adjusting for fragment duplication rates (which therefore also requires interval-level duplication rates) and anecdotally this sometimes improves results for samples with poor input, but this won't make it in the next Bioconductor release.

**Comment by chapmanb** [comment]
Luca, Markus, Gavin and Eric;
Thank you again for all this background. We've implemented an initial pass on this in the latest development version of bcbio. This moves initial steps out of CNVkit into `calculate_sv_bins` and `calculate_sv_coverage` so they can be reused by TitanCNA and PureCN implementations. What is does now is:

- Chose target regions for fine grained resolution based on the same logic as previously used -- `sv_regions` or `variant_regions` or transcripts for WGS runs.
- Use the remaining coverage regions with less fine grained bins as the antitarget regions (in CNVkit language).
- Calculate a bin size for regions based on sequencing depth, keeping it consistent across any samples shared in batches.
- Split into regions based on this bin size and target regions.
- Calculate depth in regions using mosdepth, outputting a CNVkit cnn style file (http://cnvkit.readthedocs.io/en/stable/fileformats.html?highlight=cnn#target-and-antitarget-bin-level-coverages-cnn) with regions, genes, depth and uncorrected log2 depth.

It does not yet GC and bias correct this CNN file, which is my next step. The CNVkit bias correction also does background correction based on binning and I wanted to leave that step to the downstream callers so will look into the approaches TitanCNA and PureCN use.

Hopefully this provides a standard set of ready to use regions we can build off for integration. As the next steps we'll start to tackle the caller integrations and would be happy for any feedback or suggestions. Thanks again for all this great discussion.

**Comment by lbeltrame** [comment]
I've started using TitanCNA on some pilot analyses here, so feel free to poke me should you need it when tackling TitanCNA (I admittedly didn't look at PureCNA).

**Comment by lima1** [comment]
PureCN will probably need some significant work for WGS. You'll probably want to do the normalization and segmentation with CNVkit. We originally wrote it for 2-3Mb panels and the runtime for a typical sample is about 15mins. But it can be a couple of hours for noisy whole exome samples.  Most other tools are much faster because they average coverage and allelic fractions over segments. Which we cannot do without matched normal, because we look at every single variant and check if it is likely somatic, germline or cross-sample contamination/artifact. But some kind averaging, like of neighboring common SNPs and coverages, might be necessary for WGS to get down to reasonable runtimes. 

For hybrid capture, if you provide a CNVkit style coverage file, it will look for "Background" or "Antitarget" in the gene column and will treat the bins as off vs ontarget. The GC-normalization (which is pretty much the simplified TitanCNA loess normalization) will normalize them independently if provided in a single file.  If you provide a DNAcopy style segmentation, PureCN will look for the log2 tumor vs normal coverage ratio in the coverage file and will use those in the likelihood calculation (so make sure that those are normalized).

**Comment by lima1** [comment]
One thing I forgot: SNPs outside the on-target +/- 50bp flanking regions are by default filtered out in PureCN. So depending what you define as off-target in WGS, runtime could be at least acceptable. If necessary, we could provide an option that uses segment-level instead of interval-level coverage in the likelihood model to speed-up.

**Comment by roryk** [comment]
We have support for TitanCNA and PureCN in bcbio now (though neither one is currently working due to bioconda issues!) so this should be all set.
